README.md
# cuddly-bassoon
The Ethereum Implementation Tutorial

__Why did you choose this name?__

github advised me to do so:

"Great repository names are short and memorable. Need inspiration? How about cuddly-bassoon."


__Why does this repository exist?__

I wanted to do an implementation of a simple ethereum client in easy to understand code.
And because I like to make myself challenges, to drive further work, I post these challenges and keep snapshots of their status, so that anyone can follow along, implementing their own client.
I'll track challenges/features in issues, but am listing the important ones in the readme, too.

I'll also upload test cases to see if my/your code works as intended

# Challenges

1. parse RLP encoded data
    - write a datastructure to capture RLP data
2. encode RLP trees
3. encode/decode hex values
4. hash byte arrays with keccak256
5. hash block0 (0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3)
    - find block0

# Tips and Tricks
- always have the yellow paper open
- sometimes it's helpful to look into other ethereum implementations (geth, pyethereum, cppethereum, ...)

# TODO
- expand challenge descriptions, to make it easier to follow along, maybe i'll make a blog post series out of this
    - include page references to the yellow paper
    


ChangeLog.md
# Changelog for cuddly-bassoon

## Unreleased changes


test/Spec.hs
{-# OPTIONS_GHC -F -pgmF hspec-discover #-}

-- block0 hash: 0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3


test/HexSpec.hs
module HexSpec (main, spec) where

import Test.Hspec
import CuddlyBassoon.Crypto

import CuddlyBassoon.Hex

main :: IO ()
main = hspec spec

spec :: Spec
spec = do
    describe "fromHex" $ do
        it "decodes an empty hexstring" $ do
            fromHex "" `shouldBe` Just []
        it "decodes a hexstring of 00" $ do
            fromHex "00" `shouldBe` Just [0x00]
        it "decodes a hexstring of ff" $ do
            fromHex "ff" `shouldBe` Just [0xff]
        it "decodes a hexstring of FF" $ do
            fromHex "FF" `shouldBe` Just [0xff]
        it "decodes a hexstring of 0123456789abcdef" $ do
            fromHex "0123456789abcdef" `shouldBe` Just [0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef]
    describe "fromHex failures" $ do
        it "fails to decode a hexstring of 0" $ do
            fromHex "0" `shouldBe` Nothing
        it "fails to decode a hexstring of xx" $ do
            fromHex "xx" `shouldBe` Nothing
    describe "toHex" $ do
        it "encodes an empty bytestring" $ do
            toHex [] `shouldBe` ""
        it "encodes a bytestring of [0x00]" $ do
            toHex [0x00] `shouldBe` "00"
        it "encodes a bytestring of [0xff]" $ do
            toHex [0xff] `shouldBe` "ff"
        it "encodes a bytestring of [0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef]" $ do
            toHex [0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef] `shouldBe` "0123456789abcdef"


test/RLPSpec.hs
module RLPSpec (main, spec) where

import Test.Hspec
import CuddlyBassoon.RLP

main :: IO ()
main = hspec spec

spec :: Spec
spec = do
    describe "get RLP arrays" $ do
        it "decodes an empty RLP byte array" $ do
            getRlp [0x80] `shouldBe` Just (Leaf [])
        it "decodes a scalar 0" $ do
            getRlp [0x00] `shouldBe` Just (Leaf [0x00])
        it "decodes a scalar 127" $ do
            getRlp [0x7f] `shouldBe` Just (Leaf [0x7f])
        it "decodes a scalar 128" $ do
            getRlp [0x81, 0x80] `shouldBe` Just (Leaf [0x80])
        it "decodes a scalar 255" $ do
            getRlp [0x81, 0xff] `shouldBe` Just (Leaf [0xff])
        it "decodes a scalar with 2 bytes" $ do
            getRlp [0x82, 0x00, 0x00] `shouldBe` Just (Leaf [0x00, 0x00])
        it "decodes a scalar with 55 bytes" $ do
            getRlp (0xb7 : replicate 55 0x00) `shouldBe` Just (Leaf (replicate 55 0x00))
        it "decodes a scalar with 56 bytes" $ do
            getRlp (0xb8 : 56 : replicate 56 0x00) `shouldBe` Just (Leaf (replicate 56 0x00))

    describe "get RLP array failures" $ do
        it "fails to decode an array of length 0 given length 1" $ do
            getRlp [0x81] `shouldBe` Nothing
        it "fails to decode an array of length 1 given length 2" $ do
            getRlp [0x82, 0x00] `shouldBe` Nothing
        it "fails to decode an array of length 2 given length 1" $ do
            getRlp [0x81, 0x00, 0x00] `shouldBe` Nothing
        it "fails to decode an array of length 1 with a length encoding of >55" $ do
            getRlp [0xb8, 0x01, 0x00] `shouldBe` Nothing
        it "fails to decode an array of length 56 where the length has an extra 0x00" $ do
            getRlp ([0xb9, 0x00, 56] ++ replicate 56 0x00) `shouldBe` Nothing

    describe "get RLP sequence" $ do
        it "decodes an empty RLP sequence " $ do
            getRlp [0xc0] `shouldBe` Just (Node [])
        it "decodes a RLP sequence of 1 scalar 0" $ do
            getRlp [0xc1, 0x00] `shouldBe` Just (Node [Leaf [0x00]])
        it "decodes a RLP sequence of 55 scalar 0" $ do
            getRlp (0xf7 : replicate 55 0x00) `shouldBe` Just (Node (replicate 55 (Leaf [0x00])))
        it "decodes a RLP sequence of 56 scalar 0" $ do
            getRlp (0xf8 : 56 : replicate 56 0x00) `shouldBe` Just (Node (replicate 56 (Leaf [0x00])))
        it "decodes a RLP sequence of 1 scalar with 54 bytes" $ do
            getRlp (0xf7 : 0xb6 : replicate 54 0x00) `shouldBe` Just (Node [Leaf (replicate 54 0x00)])
        it "decodes a RLP sequence of 1 scalar with 55 bytes" $ do
            getRlp (0xf8 : 56 : 0xb7 : replicate 55 0x00) `shouldBe` Just (Node [Leaf (replicate 55 0x00)])
        it "decodes a RLP sequence of 1 scalar with 56 bytes" $ do
            getRlp (0xf8 : 58 : 0xb8 : 56 : replicate 56 0x00) `shouldBe` Just (Node [Leaf (replicate 56 0x00)])
        it "decodes a RLP sequence of 1 RLP sequence of 1 scalar 0" $ do
            getRlp [0xc2, 0xc1, 0x00] `shouldBe` Just (Node [Node [Leaf [0x00]]])

    describe "get RLP sequence failures" $ do
        it "fails to decode an empty RLP sequence with extra bytes" $ do
            getRlp [0xc0, 0x00] `shouldBe` Nothing
        it "fails to decode a RLP sequence of 2 scalar 0 given length 1" $ do
            getRlp [0xc1, 0x00, 0x00] `shouldBe` Nothing
        it "fails to decode a RLP sequence of 1 scalar 0 given length 2" $ do
            getRlp [0xc2, 0x00] `shouldBe` Nothing
        it "fails to decode a RLP sequence of length 1 with a length encoding of >55" $ do
            getRlp [0xf8, 0x01, 0x00] `shouldBe` Nothing
        it "fails to decode a RLP sequence of length 56 where the length has an extra 0x00" $ do
            getRlp (0xf9 : 0x00 : 56 : 0xb7 : replicate 55 0x00) `shouldBe` Nothing

    describe "put RLP array" $ do
        it "encodes an empty RLP array" $ do
            putRlp (Leaf []) `shouldBe` [0x80]
        it "encodes a scalar 0" $ do
            putRlp (Leaf [0x00]) `shouldBe` [0x00]
        it "encodes a scalar 127" $ do
            putRlp (Leaf [0x7f]) `shouldBe` [0x7f]
        it "encodes a scalar 128" $ do
            putRlp (Leaf [0x80]) `shouldBe` [0x81, 0x80]
        it "encodes a scalar with 55 bytes" $ do
            putRlp (Leaf (replicate 55 0x00)) `shouldBe` 0xb7 : replicate 55 0x00
        it "encodes a scalar with 56 bytes" $ do
            putRlp (Leaf (replicate 56 0x00)) `shouldBe` 0xb8 : 56 : replicate 56 0x00

    describe "put RLP sequence" $ do
        it "encodes an empty RLP sequence" $ do
            putRlp (Node []) `shouldBe` [0xc0]
        it "encodes a RLP sequence of 1 scalar 0" $ do
            putRlp (Node [Leaf [0x00]]) `shouldBe` [0xc1, 0x00]
        it "encodes a RLP sequence of 55 scalar 0" $ do
            putRlp (Node (replicate 55 (Leaf [0x00]))) `shouldBe` (0xf7 : replicate 55 0x00)
        it "encodes a RLP sequence of 56 scalar 0" $ do
            putRlp (Node (replicate 56 (Leaf [0x00]))) `shouldBe` (0xf8 : 56 : replicate 56 0x00)
        it "encodes a RLP sequence of 1 scalar with 54 bytes" $ do
            putRlp (Node [Leaf (replicate 54 0x00)]) `shouldBe` (0xf7 : 0xb6 : replicate 54 0x00)
        it "encodes a RLP sequence of 1 scalar with 55 bytes" $ do
            putRlp (Node [Leaf (replicate 55 0x00)]) `shouldBe` (0xf8 : 56 : 0xb7 : replicate 55 0x00)
        it "encodes a RLP sequence of 1 scalar with 56 bytes" $ do
            putRlp (Node [Leaf (replicate 56 0x00)]) `shouldBe` (0xf8 : 58 : 0xb8 : 56 : replicate 56 0x00)
        it "encodes a RLP sequence of 1 RLP sequence of 1 scalar 0" $ do
            putRlp (Node [Node [Leaf [0x00]]]) `shouldBe` [0xc2, 0xc1, 0x00]


src/CuddlyBassoon/Hex.hs
module CuddlyBassoon.Hex where

import Data.Maybe
import GHC.Word (Word8)
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Lazy.Char8 as C

toHex :: [Word8] -> String
toHex = concatMap w
    where
        s = "0123456789abcdef"
        w ch = [s !! div x 16,s !! mod x 16]
            where x = fromEnum ch

fromHex' :: String -> [Word8]
fromHex' = fromJust . fromHex
fromHex :: String -> Maybe [Word8]
fromHex [] = Just []
fromHex (a:b:r) = do
    x <- c a
    y <- c b
    xs <- fromHex r
    return (toEnum ((x * 16) + y) : xs)
fromHex _ = Nothing


c :: Char -> Maybe Int
c '0' = Just 0
c '1' = Just 1
c '2' = Just 2
c '3' = Just 3
c '4' = Just 4
c '5' = Just 5
c '6' = Just 6
c '7' = Just 7
c '8' = Just 8
c '9' = Just 9
c 'A' = Just 10
c 'B' = Just 11
c 'C' = Just 12
c 'D' = Just 13
c 'E' = Just 14
c 'F' = Just 15
c 'a' = Just 10
c 'b' = Just 11
c 'c' = Just 12
c 'd' = Just 13
c 'e' = Just 14
c 'f' = Just 15
c _   = Nothing


src/CuddlyBassoon/Convert.hs
module CuddlyBassoon.Convert where

import GHC.Word (Word8)
import Data.Bits (Bits, (.&.), (.|.), shiftR, shiftL)

putBe :: (Integral a, Bits a) => a -> [Word8]
putBe = reverse . putLe

putLe :: (Integral a, Bits a) => a -> [Word8]
putLe 0 = []
putLe x
    | x < 256   = [fromIntegral x]
    | otherwise = (fromIntegral x .&. 0xFF) : putLe (x `shiftR` 8)

getLe :: (Integral a, Bits a) => [Word8] -> a
getLe [] = 0
getLe (x:xs) = getLe xs `shiftL` 8 .|. fromIntegral x

-- no following zeros allowed
getLeShortest :: (Integral a, Bits a) => [Word8] -> Maybe a
getLeShortest [] = Just 0
getLeShortest [0] = Nothing
getLeShortest (x:xs) = do
    b <- getLeShortest xs
    return (b `shiftL` 8 .|. fromIntegral x)

getBe :: (Integral a, Bits a) => [Word8] -> a
getBe = getLe . reverse

--no starting zeros allowed
getBeShortest :: (Integral a, Bits a) => [Word8] -> Maybe a
getBeShortest = getLeShortest . reverse


src/CuddlyBassoon/RLP/Parser.hs
module CuddlyBassoon.RLP.Parser where

import GHC.Word (Word8)
import Control.Applicative (Alternative(..))

newtype Parser a = Parser { runParser :: [Word8] -> Maybe (a, [Word8]) }

instance Functor Parser where
    fmap f (Parser p) = Parser $ \i -> case p i of
        Just (o, i') -> Just (f o, i')
        Nothing -> Nothing

instance Applicative Parser where
    pure o = Parser $ \i -> Just (o, i)
    Parser pf <*> Parser px = Parser $ \i -> case pf i of
        Nothing -> Nothing
        Just (f, i') -> case px i' of
            Nothing -> Nothing
            Just (x, i'') -> Just (f x, i'')

instance Alternative Parser where
    empty = Parser $ \i -> Nothing
    Parser p <|> Parser q = Parser $ \i -> case p i of
        Just (o, i') -> Just (o, i')
        Nothing -> q i

instance Monad Parser where
    return = pure
    Parser k >>= f = Parser $ \i -> case k i of
        Nothing -> Nothing
        Just (o, i') -> runParser (f o) i'


-- parsing:
parse :: Parser a -> [Word8] -> Maybe a
parse p i = fst <$> runParser (final p) i

final :: Parser a -> Parser a
final p = p <* eoi

-- basic parsers:
item :: Parser Word8
item = Parser $ \i -> case i of
    [] -> Nothing
    (o:i') -> Just (o, i')

eoi :: Parser ()
eoi = Parser $ \i -> case i of
    [] -> Just ((), [])
    _ -> Nothing

-- combinators:
manyN :: Integral n => Parser a -> n -> Parser [a]
manyN _ 0 = return []
manyN p n = do
    x <- p
    xs <- manyN p (n - 1)
    return (x:xs)

force :: Parser (Maybe a) -> Parser a
force (Parser p) = Parser $ \i -> case p i of
    Just ((Just o), i') -> Just (o, i')
    _ -> Nothing

bytes :: Integral a => a -> Parser [Word8]
bytes = manyN item


src/CuddlyBassoon/RLP.hs
module CuddlyBassoon.RLP (Tree(..), RLPTree, putRlp, getRlp) where

import CuddlyBassoon.Convert (putBe, getBeShortest)
import CuddlyBassoon.RLP.Parser (Parser, parse, item, force, bytes)

import Control.Applicative (empty, many)
import GHC.Word (Word8)

data Tree a = Leaf a | Node [Tree a]
    deriving (Eq, Show)

type RLPTree = Tree [Word8]

putRlp :: RLPTree -> [Word8]
putRlp (Leaf xs)
    | l == 1 && x < 128 = [x]
    | l < 56            = [128 + fromIntegral l] ++ xs
    | otherwise         = [183 + fromIntegral lb] ++ b ++ xs
    where
        l = length xs
        x = head xs
        b = putBe l
        lb = length b
putRlp (Node xs)
    | ls < 56   = [192 + fromIntegral ls] ++ s
    | otherwise = [247 + fromIntegral lb] ++ b ++ s
    where
        s = concatMap putRlp xs
        ls = length s
        b = putBe ls
        lb = length b

getRlp :: [Word8] -> Maybe RLPTree
getRlp = parse tree

length_shortest :: Word8 -> Parser Integer
length_shortest n = force (getBeShortest <$> bytes n)

trees :: Integral a => a -> Parser [RLPTree]
trees n = force (parse (many tree) <$> bytes n)

tree :: Parser RLPTree
tree = do
    i <- item
    case i of
        n | n <  0x80             -> return (Leaf [n])
        n | n >= 0x80 && n < 0xb8 -> Leaf <$> bytes (n - 0x80)
        n | n >= 0xb8 && n < 0xc0 -> do
            l <- length_shortest (n - 0xb7)
            if l < 0x38 then empty else Leaf <$> bytes l
        n | n >= 0xc0 && n < 0xf8 -> Node <$> trees (n - 0xc0)
        n | n >= 0xf8             -> do
            l <- length_shortest (n - 0xf7)
            if l < 0x38 then empty else Node <$> trees l
